modules/jdbc/src/main/java/org/testcontainers/jdbc/ConnectionUrl.java

```diff
@@ -0,0 +1,209 @@
+        /*
+        Extract from the JDBC connection URL:
+         * The database type (e.g. mysql, postgresql, ...)
+         * The docker tag, if provided.
+         * The URL query string, if provided
+       */
+        Matcher urlMatcher = Patterns.URL_MATCHING_PATTERN.matcher(this.getUrl());
+        if (!urlMatcher.matches()) {
+            //Try for Oracle pattern
+            urlMatcher = Patterns.ORACLE_URL_MATCHING_PATTERN.matcher(this.getUrl());
+            if (!urlMatcher.matches()) {
+                throw new IllegalArgumentException("JDBC URL matches jdbc:tc: prefix but the database or tag name could not be identified");
+            }
+        }
+        databaseType = urlMatcher.group(1);
+
+        imageTag = Optional.ofNullable(urlMatcher.group(3)).orElse("latest");
+
+        //String like hostname:port/database name, which may vary based on target database.
+        //Clients can further parse it as needed.
+        dbHostString = urlMatcher.group(4);
+
+        //In case it matches to the default pattern
+        Matcher dbInstanceMatcher = Patterns.DB_INSTANCE_MATCHING_PATTERN.matcher(dbHostString);
+        if (dbInstanceMatcher.matches()) {
+            databaseHost = Optional.of(dbInstanceMatcher.group(1));
+            databasePort = Optional.ofNullable(dbInstanceMatcher.group(3)).map(value -> Integer.valueOf(value));
+            databaseName = Optional.of(dbInstanceMatcher.group(4));
+        }
+
+        queryString = Optional.ofNullable(urlMatcher.group(5));
+        getQueryParameters();
+
+        Matcher matcher = Patterns.INITSCRIPT_MATCHING_PATTERN.matcher(this.getUrl());
+        if (matcher.matches()) {
+            initScriptPath = Optional.ofNullable(matcher.group(2));
+        }
+
+        Matcher funcMatcher = Patterns.INITFUNCTION_MATCHING_PATTERN.matcher(this.getUrl());
+        if (funcMatcher.matches()) {
+            initFunction = Optional.of(new InitFunctionDef(funcMatcher.group(2), funcMatcher.group(4)));
+        }
+
+        Matcher daemonMatcher = Patterns.DAEMON_MATCHING_PATTERN.matcher(this.getUrl());
+        inDaemonMode = daemonMatcher.matches() ? Boolean.parseBoolean(daemonMatcher.group(2)) : false;
+
+    }
+
+    /**
+     * Get the TestContainers Parameters such as Init Function, Init Script path etc.
+     *
+     * @return {@link Map}
+     */
+    public Map<String, String> getContainerParameters() {
+
+        Map<String, String> results = new HashMap<>();
+
+        Matcher matcher = Patterns.TC_PARAM_MATCHING_PATTERN.matcher(this.getUrl());
+        while (matcher.find()) {
+            String key = matcher.group(1);
+            String value = matcher.group(2);
+            results.put(key, value);
+        }
+
+        return results;
+    }
+
+    /**
+     * Get all Query paramters specified in the Connection URL after ?. This also includes TestContainers parameters.
+     *
+     * @return {@link Map}
+     */
+    public Map<String, String> getQueryParameters() {
```

Comment:
It looks like this method mutates the `queryString` field as a side effect of constructing the Map. If someone calls `getQueryString` too early, they'll get nothing. 

Please could we avoid this? Maybe we could make all fields be final (which a static factory method and maybe Lombok could help with)?

Reply:
I have modified the query and container parameter extraction method to make them immutable. Now, `queryParameters` and `containerParameters` have getters that returns unmodifiable maps and `queryString` is instantiated only once during initial parsing. Please check.